osu.Game/Screens/Select/FilterQueryParser.cs

```diff
@@ -34,6 +36,73 @@ internal static void ApplyQueries(FilterCriteria criteria, string query)
             criteria.SearchText = query;
         }
 
+        /// <summary>
+        /// Extracts and returns the difficulty search text between square brackets.
+        /// </summary>
+        /// <param name="query">The search query. The difficulty search text will be removed from the query.</param>
+        /// <returns>The difficulty search text (without the square brackets).</returns>
+        private static string extractDifficultySearchText(ref string query)
+        {
+            var openingBracketIndexes = new List<int>();
+            var closingBracketIndexes = new List<int>();
+
+            populateIndexLists(ref query);
+
+            return performExtraction(ref query);
+
+            void populateIndexLists(ref string query)
+            {
+                bool currentlyBetweenBrackets = false;
+
+                for (int i = 0; i < query.Length; i++)
+                {
+                    switch (query[i])
+                    {
+                        case '[' when !currentlyBetweenBrackets && (i == 0 || query[i - 1] == ' '):
+                            currentlyBetweenBrackets = true;
+                            openingBracketIndexes.Add(i + 1);
+                            break;
+
+                        case ']' when currentlyBetweenBrackets && (i == query.Length - 1 || query[i + 1] == ' '):
+                            currentlyBetweenBrackets = false;
+                            closingBracketIndexes.Add(i);
+                            break;
+                    }
+                }
+
+                if (currentlyBetweenBrackets)
+                {
+                    // If there is no "]" closing the current difficulty search query, append it.
+                    query += ']';
```

Comment:
this is writing to the `ref` variable. seems super wrong

Reply:
The reason for this was that, if there was no closing bracket, it needed to be added so `performExtraction()` can process it correctly. I changed `performExtraction()` so it can process it regardless, so I removed the `ref` parameter entirely.